addons/account/models/partner.py

```diff
@@ -248,15 +230,22 @@ def _get_fiscal_position(self, partner, delivery=None):
         if manual_fiscal_position:
             return manual_fiscal_position
 
-        # First search only matching VAT positions
-        vat_valid = self._get_vat_valid(delivery, company)
-        fp = self._get_fpos_by_region(delivery.country_id.id, delivery.state_id.id, delivery.zip, vat_valid)
-
-        # Then if VAT required found no match, try positions that do not require it
-        if not fp and vat_valid:
-            fp = self._get_fpos_by_region(delivery.country_id.id, delivery.state_id.id, delivery.zip, False)
-
-        return fp or self.env['account.fiscal.position']
+        # Search for a an auto applied fiscal position matching the partner
+        ranking_subfunctions = self._get_fpos_ranking_functions(delivery)
+        def ranking_function(fpos):
+            return tuple(rank[1](fpos) for rank in ranking_subfunctions)
+
+        return max(
+            filter(lambda x: all(x[1]), [
+                (fpos, ranking_function(fpos))
+                for fpos in self.search([
+                    *self._check_company_domain(self.env.company),
+                    ('auto_apply', '=', True),
+                ])
```

Comment:
If we wanted to get almost completely rid of any `search` to find fiscal positions in loops, we could store (in memory) the fiscal positions of a company on a non stored (in db) field of the company.

Probably over engineering for now though.